freeipa/src/main/java/com/sequenceiq/freeipa/service/stack/FreeIpaHealthDetailsService.java

```diff
@@ -53,35 +58,49 @@
 
     public HealthDetailsFreeIpaResponse getHealthDetails(String environmentCrn, String accountId) {
         Stack stack = stackService.getByEnvironmentCrnAndAccountIdWithLists(environmentCrn, accountId);
-        InstanceMetaData master = findMaster(stack);
-        Optional<RPCResponse<Boolean>> rpcResponse = Optional.empty();
-        try {
-            rpcResponse = Optional.ofNullable(checkFreeIpaHealth(stack, master));
-        } catch (FreeIpaClientException e) {
-            LOGGER.error("Unable to check the health of FreeIPA.", e);
+        List<InstanceMetaData> instances = stack.getAllInstanceMetaDataList();
+        HealthDetailsFreeIpaResponse response = new HealthDetailsFreeIpaResponse();
+
+        for (int idx = 0; idx < instances.size(); idx++) {
+            InstanceMetaData instance =  instances.get(idx);
+            if (isInstanceAvailable(instance, response)) {
+                try {
+                    RPCResponse<Boolean> rpcResponse = checkFreeIpaHealth(stack,  instance);
+                    parseMessages(rpcResponse, response);
+                } catch (FreeIpaClientException e) {
+                    NodeHealthDetails nodeResponse = new NodeHealthDetails();
+                    response.addNodeHealthDetailsFreeIpaResponses(nodeResponse);
+                    nodeResponse.setName(instance.getDiscoveryFQDN());
+                    nodeResponse.setStatus(InstanceStatus.UNREACHABLE);
+                    nodeResponse.addIssue(e.getLocalizedMessage());
+                    LOGGER.error(String.format("Unable to check the health of FreeIPA instance: %s", instance.getInstanceId()), e);
+                }
+            }
         }
-        return createResponse(stack, rpcResponse, master);
+        return updateResponse(stack, response);
     }
 
-    private HealthDetailsFreeIpaResponse createResponse(Stack stack, Optional<RPCResponse<Boolean>> rpcResponse, InstanceMetaData master) {
-        HealthDetailsFreeIpaResponse response = new HealthDetailsFreeIpaResponse();
-        response.setEnvironmentCrn(stack.getEnvironmentCrn());
-        response.setCrn(stack.getResourceCrn());
-        if (rpcResponse.isPresent()) {
-            response.setName((String) rpcResponse.get().getValue());
-            parseMessages(rpcResponse.get(), response);
-            if (isOverallHealthy(response)) {
-                response.setStatus(DetailedStackStatus.PROVISIONED.getStatus());
-            } else {
-                response.setStatus(DetailedStackStatus.UNHEALTHY.getStatus());
-            }
+    private boolean isInstanceAvailable(InstanceMetaData instance, HealthDetailsFreeIpaResponse response) {
```

Comment:
the method name doesn't reflect what it does. I think `instance.isAvailable()` should be in the caller if, and the else branch from there should call the body of this method (wo the if here oc)

Reply:
changed